File: 0_vectors/C/vector.c
/*
Heru Handika
6 September 2020
Vector operation includes:
Addition
Substraction
Multiplication
Summation
Mean
Dot product
Sum of square
Magnitude
Distance
*/
#include "vector.h"

void print_vectors(double vec[], size_t vec_size)
{
    for (size_t i = 0; i < vec_size; i++)
    {
        printf("%.2f ", vec[i]);
    }
    printf("\n");
}

double allocate_mem(double **arr, size_t size)
{
    *arr = (double *)malloc(size * sizeof(double));
    if (*arr == 0)
    {
        printf("\x1b[0;31Error: \x1b[0mFailed to allocate memory\n");
        exit(EXIT_FAILURE);
    }
    return 0;
}

// Avoid double free memory.
void free_mem(void **pointer)
{
    if (pointer != NULL && *pointer != NULL)
    {
        free(*pointer);
        *pointer = NULL;
    }
}

double *vec_ones(size_t arr_size)
{
    double *presult = NULL;
    allocate_mem(&presult, arr_size);
    for (size_t i = 0; i < arr_size; i++)
    {
        presult[i] = 1;
    }
    return presult;
}

double *add_vectors(double x[], double y[], size_t arr_size)
{
    double *presult = NULL;
    allocate_mem(&presult, arr_size);
    for (size_t i = 0; i < arr_size; i++)
    {
        presult[i] = x[i] + y[i];
    }
    return presult;
}

double *substract_vectors(double x[], double y[], size_t arr_size)
{
    double *presult = NULL;
    allocate_mem(&presult, arr_size);
    for (size_t i = 0; i < arr_size; i++)
    {
        presult[i] = x[i] - y[i];
    }
    return presult;
}

double *multiply_vectors(double x[], double y[], size_t arr_size)
{
    double *presult = NULL;
    allocate_mem(&presult, arr_size);
    for (size_t i = 0; i < arr_size; i++)
    {
        presult[i] = x[i] * y[i];
    }
    return presult;
}

double sum(double vec[], size_t arr_size)
{
    double sum = 0;
    size_t i = 0;

    while (i < arr_size)
    {
        sum += vec[i];
        i++;
    }
    return sum;
}

double *cumsum(double vec[], size_t arr_size)
{
    double *res = NULL;
    allocate_mem(&res, arr_size);

    double sum = 0;
    for (size_t i = 0; i < arr_size; i++)
    {
        sum += vec[i];
        res[i] = sum;
    }

    return res;
}

double mean(double sum, size_t arr_size)
{
    return (double)sum / arr_size;
}

// Dot product will be use to compute sum of square
double dot_product(double x[], double y[], size_t arr_size)
{
    double sum = 0.0;
    for (size_t i = 0; i < arr_size; i++)
    {
        sum += x[i] * y[i];
    }
    return sum;
}

double sum_of_square(double a[], size_t arr_size)
{
    return dot_product(a, a, arr_size);
}

double magnitude(double a[], size_t arr_size)
{
    return sqrt(sum_of_square(a, arr_size));
}

double distance(double x[], double y[], size_t arr_size)
{
    double *vec_subs = substract_vectors(x, y, arr_size);
    double mag = magnitude(vec_subs, arr_size);
    FREE_MEM(vec_subs);
    return mag;
}

File: 0_vectors/rust/vector.rs
/*
Heru Handika
19 September 2020
Vector addition and multiplication in rust
*/

use std::f64;

fn main() {
    let x: Vec<i32> = vec![1, 2, 3, 4, 5];
    let y: Vec<i32> = vec![10, 20, 30, 40, 50];

    println!("Vector ones: {:?}",
        vector_ones(5));

    println!("Vector addition: {:?}", 
        add_two_vectors(&x, &y));

    println!("Vector multiplication: {:?}", 
        multiply_two_vectors(&x, &y));
    
    println!("Vector substraction: {:?}",
        substract_two_vectors(&x, &y));

    println!("Sum x: {}", sum_vectors(&x));
    println!("Sum y: {}", sum_vectors(&y));
    println!("Mean x: {:.2}", mean_vectors(&x));
    println!("Mean y: {:.2}", mean_vectors(&y));

    println!("Dot product x & y: {:.2}", 
        calculate_dot_product(&x, &y));
    println!("Sum of square x: {:.2}", 
        calculate_sum_of_square(&x));
    println!("Sum of square y: {:.2}", 
        calculate_sum_of_square(&y));
    println!("Magnitude x: {:.4}", 
        compute_magnitude(&x));
    println!("X & Y distance: {:.4}", 
        compute_vector_distance(&x, &y));
}

fn vector_ones(vec_size: usize) -> Vec<i32> {
    let mut vec_ones: Vec<i32> = Vec::with_capacity(vec_size);
    for _ in 0..vec_size {
        vec_ones.push(1);
    }
    vec_ones
}

fn add_two_vectors(x: &Vec<i32>, y: &Vec<i32>) -> Vec<i32>{
    let mut result = Vec::new();

    for (&i, &j) in x.iter().zip(y.iter()) {
        result.push(i + j);
    }
    result
}

fn substract_two_vectors(x: &Vec<i32>, y: &Vec<i32>) -> Vec<i32> {
    let mut result = Vec::new();

    for (&i, &j) in x.iter().zip(y.iter()) {
        result.push(i - j);
    }
    result
}

fn multiply_two_vectors(x: &Vec<i32>, y: &Vec<i32>) -> Vec<i32>{
    let mut result = Vec::new();
    
    for (&i, &j) in x.iter().zip(y.iter()) {
        result.push(i * j);
    }
    result
}

fn sum_vectors(x: &Vec<i32>) -> i32 {
    let mut sum = 0;
    for i in x.iter() {
        sum += i;
    }
    sum
}

fn mean_vectors(vec: &Vec<i32>) -> f64 {
    let n = vec.len();
    let mean = sum_vectors(vec) as f64 / n as f64;

    mean
}

// Dot product is the measure of how far two vectors are
// Will be used to calculate sum of squares
fn calculate_dot_product(x: &Vec<i32>, y: &Vec<i32>) -> f64 {
    let mut sum: f64 = 0.0;
    for (&i, &j) in x.iter().zip(y.iter()) {
        sum += i as f64 * j as f64;
    }
    sum
}

fn calculate_sum_of_square(vec: &Vec<i32>) -> f64 {
    calculate_dot_product(vec, vec)
}

fn compute_magnitude(vec: &Vec<i32>) -> f64 {
    calculate_sum_of_square(&vec).sqrt()
}

fn compute_vector_distance(x: &Vec<i32>, y: &Vec<i32>) -> f64 {
    let vec: Vec<i32> = substract_two_vectors(&x, &y);
    let dist = compute_magnitude(&vec);
    dist
}